File: tests/cpp_tests/test_Operation.cpp

#include "../cpp_compiler/Measure.hpp"
#include "../cpp_compiler/Rotation.hpp"
#include <vector>
#include <iostream>

using namespace std;


int Test_isCommute()
{
    
    
    vector<char> x = {'x'};
    vector<int> zero = {0} ;
    Rotation X    = Rotation(0,x,zero);
    vector<char> y = {'y'};
    Rotation Y    = Rotation(0,y,zero);
    vector<char> z = {'z'};
    Rotation Z    = Rotation(0,z,zero);
    Rotation XY   = Rotation(0,vector<char> {'x','y'},{0,1}); 
    vector<int> one = {1};
    Rotation IZ   = Rotation(0,z,one);
    Rotation ZYZ  = Rotation(0,vector<char> {'z','y','z'},{0,1,2});
    Rotation XZY  = Rotation(0,vector<char> {'x','z','y'},{0,1,2});
 
    Measure MX   = Measure(0,x,zero);
    Measure MY   = Measure(0,y,zero);
    Measure MZ   = Measure(0,z,zero);
    Measure MXY  = Measure(0,vector<char> {'x','y'},{0,1});
    Measure MXZY = Measure(0,vector<char> {'x','z','y'},{0,1,2});

    vector< vector<Operation> > cases = {{X,Y,Z,XY,X,Y,Z,XY,ZYZ,X,Y,Z,XY,X,Y,Z,XY,ZYZ},{MX,MY,MZ,MXY,MY,MZ,MX,IZ,MXZY,X,Y,Z,XY,Y,Z,X,IZ,XZY}};
    vector<bool> expected = {true,true,true,true,false,false,false,false,false,true,true,true,true,false,false,false,false,false};

    for(int idx = 0 ; idx < cases[0].size() ; ++idx){
        if (expected[idx] != cases[0][idx].isCommute(cases[1][idx])){
            cout << cases[0][idx].Print() ;
            cout << cases[1][idx].Print() ;
            cout << "case " << idx << "failed\n" ;
            return 1;
        }
    }
    
    cout << "isCommute passed\n" ;
    return 0;
}

int test_isIdentity(){
    vector<char> x = {'x'};
    vector<int> zero = {0};
    Rotation XIII = Rotation(0,x,zero);
    vector<char> emptyc = {};
    vector<int> emptyi = {};
    Rotation III  = Rotation(0,emptyc,emptyi);

    vector<Rotation> cases    = {XIII,III} ;
    vector<bool> expected = {false,true} ;

    for(int idx = 0 ; idx < expected.size() ; ++idx){
        if (expected[idx] != cases[idx].isIdentity()){
            cout << "case " << idx << "failed\n" ;
            return 1;
        }
    }
        
    cout << "isIdentity passed\n" ;
    return 0;

}

int test_isSingleQubit(){
    vector<char> x = {'x'};
    vector<char> z = {'z'};
    vector<char> y = {'y'};
    vector<int> zero = {0};
    vector<int> one = {1};
    vector<int> three = {3};
    Rotation XIII = Rotation(0,x,zero);
    Rotation IZII = Rotation(0,z,one);
    Rotation IIIY = Rotation(0,y,three);
    Rotation YXII = Rotation(0,vector<char> {'y','x'},{0,1});
    Rotation YXZI = Rotation(0,vector<char> {'y','x','z'},{0,1,2});

    vector<Rotation> cases    = {XIII,IZII,IIIY,YXII,YXZI} ;
    vector<bool> expected = {true,true,true,false,false} ; 

    for(int idx = 0 ; idx < expected.size() ; ++idx){
        if (expected[idx] != cases[idx].isSingleQubit()){
            cout << "case " << idx << "failed\n" ;
            return 1;
        }
    }
        
    cout << "isSingleQubit passed\n" ;
    return 0;
}


int main(){
    if (Rotation(0,"0","1").xBasis.size() < 3){
        cout << "Error: Test_Operation needs to be compile for at least 3 Qubits\n" ;
        return 1;
    }
    vector<int> results = {Test_isCommute(),test_isIdentity(), test_isSingleQubit()};
    int count = 0 ;
    int nTests = results.size();
    for (int idx = 0 ; idx < nTests ; ++idx) {
        count = count + results[idx];
    }
    cout << "------------------------------------------\n";
    cout << "Tests passed in test_Operation: " << nTests-count << "/" << nTests <<"\n";
    cout << "------------------------------------------\n";
    return 0;
}

File: tests/python_tests/test_Operation.py
from src.python_wrapper.Rotation import *
from src.python_wrapper.Measure import *
import pytest
from math import pi

@pytest.fixture 
def op():
    return Operation(3, [], [])

def test_is_commute(op):
    X    = Rotation(1,0,['x'],[0])
    Y    = Rotation(1,0,['y'],[0])
    Z    = Rotation(1,0,['z'],[0])
    XY   = Rotation(2,0,['x','y'],[0,1]) 
    IZ   = Rotation(2,0,['z'],[1])
    ZYZ  = Rotation(3,0,['z','y','z'],[0,1,2])
    XZY  = Rotation(3,0,['x','z','y'],[0,1,2])
 
    MX   = Measure(1,True,['x'],[0])
    MY   = Measure(1,True,['y'],[0])
    MZ   = Measure(1,True,['z'],[0])
    MIZ  = Measure(2,True,['z'],[1])
    MXY  = Measure(2,True,['x','y'],[0,1])
    MXZY = Measure(3,True,['x','z','y'],[0,1,2])

    cases = [
        [X,Y,Z,XY,X,Y,Z,XY,ZYZ,X,Y,Z,XY,X,Y,Z,XY,ZYZ],
        [MX,MY,MZ,MXY,MY,MZ,MX,IZ,MXZY,X,Y,Z,XY,Y,Z,X,IZ,XZY]
    ]

    expected = [True,True,True,True,False,False,False,False,False,True,True,True,True,False,False,False,False,False]
    results  = []

    for each_pair in zip(cases[0],cases[1]):
        results.append(each_pair[0].is_commute(each_pair[1]))

    assert expected == results


def test_is_identity(op):
    XIII = Rotation(4,0,['x'],[0])
    III  = Rotation(3,0)

    cases    = [XIII,III]
    expected = [False,True]
    results  = []
    for each in cases:
        results.append(each.is_identity())
    
    assert expected == results


def test_is_single_qubit(op):

    XIII = Rotation(4,0,['x'],[0])
    IZII = Rotation(4,0,['z'],[1])
    IIIY = Rotation(4,0,['y'],[3])
    YXII = Rotation(4,0,['y','x'],[0,1])
    YXZI = Rotation(4,0,['y','x','z'],[0,1,2])

    cases = [
        XIII,IZII,IIIY,YXII,YXZI
    ]
    expected = [True,True,True,False,False]
    results  = []

    for each in cases:
        results.append(each.is_single_qubit())

    assert expected == results

    
